### PR TITLE
KON-188 Add Architecture Tester Project

### PR DESCRIPTION
**Pull request by nataliapeterwas**
Continue to https://github.com/LemonAppDev/konsist/pull/249

### CODE REVIEW

File: lib/src/architectureIntegrationTest/kotlin/com/lemonappdev/konsist/architecture3/Architecture3Test.kt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import com.lemonappdev.konsist.core.exception.KoPreconditionFailedException
+import com.lemonappdev.konsist.core.filesystem.PathProvider
+import com.lemonappdev.konsist.core.verify.assert
+import org.amshove.kluent.shouldThrow
+import org.amshove.kluent.withMessage
+import org.junit.jupiter.api.Test
+
+class Architecture3Test {
+    private val rootPath = PathProvider.getInstance().rootProjectPath
+
```
Comment (nataliapeterwas) [inline lib/src/architectureIntegrationTest/kotlin/com/lemonappdev/konsist/architecture3/Architecture3Test.kt:15]: nie wiem czy ten `architecure3` ma teraz sens, gdy dodałam nowy package `circulardependency`. Ewentualnie można zostawić te dwa ostatnie testy?

Comment (igorwojda) [reply lib/src/architectureIntegrationTest/kotlin/com/lemonappdev/konsist/architecture3/Architecture3Test.kt:15]: Remove all of this

File: lib/src/architectureIntegrationTest/kotlin/com/lemonappdev/konsist/architecture1/Architecture1Test.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import org.amshove.kluent.withMessage
+import org.junit.jupiter.api.Test
+
+class Architecture1Test {
+    @Test
+    fun `passes when dependency is set that layers are independent`() {
+        // given
+        val domain = Layer("Domain", "com.lemonappdev.konsist.architecture1.project.domain..")
+        val presentation = Layer("Presentation", "com.lemonappdev.konsist.architecture1.project.presentation..")
+        val koArchitecture = Konsist
```
Comment (igorwojda) [inline lib/src/architectureIntegrationTest/kotlin/com/lemonappdev/konsist/architecture1/Architecture1Test.kt:17]: Add empty line after layers definition (all tests)

File: lib/src/main/kotlin/com/lemonappdev/konsist/core/architecture/KoArchitectureImpl.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        } else {
+            val lists = layersToCheck.map {
+                checkCircularDependenciesHelper(nodeLayer, it, alreadyChecked + layerToCheck, potentialCircular + layerToCheck)
+            }
+
+            lists.firstOrNull { it.last() == null } ?: emptyList()
+        }
+    }
+
+    private fun checkIfLayerIsDependentOnItself(layer: Layer, vararg addedLayers: Layer) {
```
Comment (igorwojda) [inline lib/src/main/kotlin/com/lemonappdev/konsist/core/architecture/KoArchitectureImpl.kt:79]: Parameter `addedlayera` should be called `layers` (same for other parameters in this class)

File: lib/src/main/kotlin/com/lemonappdev/konsist/core/architecture/KoArchitectureImpl.kt [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        if (addedLayers.any { it == layer }) {
+            throw KoPreconditionFailedException("Layer: $layer cannot be dependent on itself.")
+        }
+    }
+
+    private fun checkIfLayerIsAddToArchitecture(layer: Layer, addedLayers: List<Layer>? = null) {
+        addedLayers?.let {
+            if (addedLayers.any { layer -> !allLayers.contains(layer) }) {
+                throw KoPreconditionFailedException("Layers: $it is not add to the architecture.")
```
Comment (igorwojda) [inline lib/src/main/kotlin/com/lemonappdev/konsist/core/architecture/KoArchitectureImpl.kt:88]: This looks incorrect when there is one layer that violates this conditions.

Let's change a format to:
"throw KoPreconditionException("Layers not added to the architecture: /n LIST OF LAYER SEPARATED BY NEW LINE")

(same for below error)

File: lib/src/architectureIntegrationTest/kotlin/com/lemonappdev/konsist/circulardependency/CircularDependencyTest.kt [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            Illegal circular dependencies:
+            Layer(name=layer3, isDefinedBy=layer3) -->
+            Layer(name=layer1, isDefinedBy=layer1) -->
+            Layer(name=layer2, isDefinedBy=layer2) -->
+            Layer(name=layer3, isDefinedBy=layer3).
+        """.trimIndent()
+    }
+
+    @Test
+    fun `complex circular dependency 1`() {
```
Comment (igorwojda) [inline lib/src/architectureIntegrationTest/kotlin/com/lemonappdev/konsist/circulardependency/CircularDependencyTest.kt:65]: 1. Let's name these tests "circular dependency 1...5" 
2. Let's move them to separate packages like architecture1..5

**Review by igorwojda — CHANGES_REQUESTED** [review]

**Comment by igorwojda** [comment]
Throw an exception when deps relations are duplicated eg.

```kotlin
.addDependencies {
                domain.notDependOnAnyLayer()
                domain.notDependOnAnyLayer()
                presentation.notDependOnAnyLayer()
            }
```


```kotlin
.addDependencies {
                domain.notDependOnAnyLayer()
                presentation.notDependOnAnyLayer()
                presentation.notDependOnAnyLayer()
            }
```

File: lib/src/integrationTest/kotlin/com/lemonappdev/konsist/core/verify/common/CommonTest.kt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import com.lemonappdev.konsist.TestSnippetProvider
+import com.lemonappdev.konsist.api.Konsist
+import com.lemonappdev.konsist.core.architecture.Layer
+import com.lemonappdev.konsist.core.verify.assert
+import org.amshove.kluent.shouldContain
+import org.junit.jupiter.api.Test
+
+class CommonTest {
+    @Test
+    fun `declaration-assert-test-method-name`() {
```
[the remark this answers is not in the document]
Comment (nataliapeterwas) [reply lib/src/integrationTest/kotlin/com/lemonappdev/konsist/core/verify/common/CommonTest.kt:12]: This is a convention that test name is the same as snippet file name.  We use it in all integrationTest. Should I change it here?

**Comment by igorwojda** [comment]
```kotlin
// API 1
        Konsist
            .scopeFromProject()
            .architecture {
                val domain = Layer("Domain", "com.lemonappdev.konsist.architecture2.project.domain..")
                val presentation = Layer("Presentation", "com.lemonappdev.konsist.architecture2.project.presentation..")

                domain.notDependOnAnyLayer()
                presentation.dependsOn(domain)
            }.assert()


        // Optional API 2
        val koArchitecture = architecture {
            val domain = Layer("Domain", "com.lemonappdev.konsist.architecture2.project.domain..")
            val presentation = Layer("Presentation", "com.lemonappdev.konsist.architecture2.project.presentation..")

            domain.notDependOnAnyLayer()
            presentation.dependsOn(domain)
        }

        Konsist
            .scopeFromProject()
            .architecture(koArchitecture)
            .assert()
```

**Comment by nataliapeterwas** [comment]
Pasted Graphic 1

1. Maybe we should make separate packages in `konsistArchitectureApiTest`, eg. `assertarchitecture` containg all `architecture` packages and `dependencyrules` containing `DependencyRulesTest` and `circulardependency` packages?
2. What name to rename "konsistArchitectureApiTest" to?

**Comment by igorwojda** [comment]
1. Rename `konsistArchitectureApiTest` to `apiTest ` 
2. Move all architecture-related stuff to the `com.lemonappdev.konsist.architecture` package
3. Latter move all snippets to `com.lemonappdev.konsist.declaration` package

**Comment by igorwojda** [comment]
1. Method name is incorrect "Assert 'invokeMethod' has failed. Invalid dependencies (1):"
(mango https://github.com/LemonAppDev/mango/tree/upd-konsist, `ArchitectureKonsistTest`)
a. `assertArchitecture {}`
a. `assertArchitecture ()`

2. Exception when adding the next layer with the same `name` or the same `definedBy`

3. Improve error messages

```
Domain depends on nothing assertion failure:
file
file
...
```

```
data depends on domain assertion failure:
file
file
...

**Comment by igorwojda** [comment]
1. "Layer Domain was previously set as depend on nothing, so it cannot be depend on Data layer." -> "...cannot depend on Data layer."
2. Change error msg "Assert 'verify layer dependencies' has failed. Invalid dependencies:" -> "'verify layer dependencies' test has failed. Invalid dependencies:"

File: .github/workflows/check.yml [hunk 7]
```diff
@@ -72,6 +72,27 @@ jobs:
           name: integration-test-report
           path: ./lib/build/test-results/integrationTest/
 
+  konsist-architecture-api-test:
```
Comment (igorwojda) [inline .github/workflows/check.yml:75]: change to `api-test`

**Review by igorwojda — APPROVED** [review]

**Pull request merged**